test/extended/kubevirt/util.go

```diff
@@ -353,23 +363,117 @@ func InKubeVirtClusterContext(oc *exutil.CLI, body func()) {
+		func() {
+			BeforeEach(func() {
+				setMgmtFramework(f)
+				expectNoError(migrateWorkers(f))
+			})
+			body()
+		})
+}
+
 func setMgmtFramework(mgmtFramework *e2e.Framework) *exutil.CLI {
 	_, hcpNamespace, err := exutil.GetHypershiftManagementClusterConfigAndNamespace()
 	Expect(err).NotTo(HaveOccurred())
 
 	oc := exutil.NewHypershiftManagementCLI(hcpNamespace).AsAdmin()
 
-	mgmtClientSet := oc.KubeClient()
 	mgmtFramework.Namespace = &corev1.Namespace{
 		ObjectMeta: metav1.ObjectMeta{
 			Name: hcpNamespace,
 		},
 	}
-	mgmtFramework.ClientSet = mgmtClientSet
+	mgmtFramework.ClientSet = oc.KubeClient()
+	mgmtFramework.DynamicClient = oc.DynamicClient()
 
 	return oc
 }
 
+func virtualMachineInstanceClient(dc dynamic.Interface) dynamic.NamespaceableResourceInterface {
+	return dc.Resource(vmiGVR)
+}
+
+func virtualMachineInstanceMigrationClient(dc dynamic.Interface) dynamic.NamespaceableResourceInterface {
+	return dc.Resource(vmimGVR)
+}
+
+func composeVMIM(vmiNamespace, vmiName string) *unstructured.Unstructured {
+	vmim := unstructured.Unstructured{}
+	vmim.SetGroupVersionKind(vmimGVK)
+	vmim.SetNamespace(vmiNamespace)
+	vmim.SetGenerateName(vmiName + "-")
+	vmim.UnstructuredContent()["spec"] = map[string]interface{}{
+		"vmiName": vmiName,
+	}
+	return &vmim
+}
+
+func migrateWorkers(f *e2e.Framework) error {
+	By("migrating hosted cluster workers")
+	vmiClient := virtualMachineInstanceClient(f.DynamicClient)
+	vmiList, err := vmiClient.Namespace(f.Namespace.Name).List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed retrieving workers VMs at namespace %s: %v", f.Namespace.Name, err)
+	}
+
+	vmimClient := virtualMachineInstanceMigrationClient(f.DynamicClient)
+	vmims := []*unstructured.Unstructured{}
+	for _, vmi := range vmiList.Items {
+		vmim := composeVMIM(vmi.GetNamespace(), vmi.GetName())
+		if vmim, err = vmimClient.Namespace(f.Namespace.Name).Create(context.Background(), vmim, metav1.CreateOptions{}); err != nil {
+			return fmt.Errorf("failed creating vmim %s/%s: %v", vmi.GetNamespace(), vmi.GetName(), err)
+		}
+		vmims = append(vmims, vmim)
+	}
+	for _, vmim := range vmims {
+		phase := ""
+		err := wait.PollImmediate(5*time.Second, 5*time.Minute, func() (done bool, err error) {
+			vmim, err = vmimClient.Namespace(f.Namespace.Name).Get(context.Background(), vmim.GetName(), metav1.GetOptions{})
+			if err != nil {
+				return false, err
+			}
+			statusIface, ok := vmim.UnstructuredContent()["status"]
+			if !ok {
+				return false, nil
+			}
+			status, ok := statusIface.(map[string]interface{})
+			if !ok {
+				return false, fmt.Errorf("bad vmim.status: %+v", statusIface)
+			}
+			phaseIface, ok := status["phase"]
+			if !ok {
+				return false, nil
+			}
+			phase, ok = phaseIface.(string)
+			if !ok {
+				return false, fmt.Errorf("bad vmim.status.phase: %+v", phaseIface)
+			}
+			return phase == "Succeeded" || phase == "Failed", nil
```

Comment:
Before returning here, i think it would be helpful to print out the phase so we'd know if the vmim was still somehow in progress.

I think it would also be helpful to think about what kind of debug information would be helpful in the event that the migration does not succeed. We'd probably want to know information about the src and target pods, as well as the VMI and VMIM. Maybe the resources could be dumped into the test log?

Reply:
Target pods has to be present at artifacts since they are normal pods, we can dump VMIs and VMIMs

Also kubevirt pods and events should be at artifacts too.